llama : add grammar-based sampling

ejones opened the pull request:
EDITED after updates

Inspired by #1397 and [grantslatton's CFG work](https://github.com/grantslatton/llama.cpp/commit/007e26a99d485007f724957fa8545331ab8d50c3), this adds an API that takes a serialized context-free grammar to guide and constrain sampling. Also adds a sample Backus-Naur form (BNF)-like syntax in `main` for specifying a grammar for generations.

## Testing

(M2 Max, 30B)

<details>
<summary>Chess</summary>

```
 % ./main -m $LLAMA_30B_Q4_0 -n 32 -p $'A good game:\n\n' --grammar-file grammars/chess.gbnf
main: build = 674 (e550234)
main: seed  = 1688014137
llama.cpp: loading model from /Users/evan/llama-models/30B/ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 6656
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 52
llama_model_load_internal: n_layer    = 60
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 17920
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 30B
llama_model_load_internal: ggml ctx size =    0.13 MB
llama_model_load_internal: mem required  = 19756.66 MB (+ 3124.00 MB per state)
.
llama_init_from_file: kv self size  =  780.00 MB

system_info: n_threads = 8 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | 
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.800000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 512, n_batch = 512, n_predict = 32, n_keep = 0


main: grammar:
root ::= [1] [.] [ ] move [ ] move [<U+000A>] root_4 
move ::= move_5 move_9 
root_2 ::= [1-9] root_3 [.] [ ] move [ ] move [<U+000A>] 
root_3 ::= [0-9] | 
root_4 ::= root_2 root_4 | root_2 
move_5 ::= pawn | nonpawn | castle 
pawn ::= pawn_14 [a-h] [1-8] pawn_16 
nonpawn ::= [NBKQR] nonpawn_10 nonpawn_11 nonpawn_12 [a-h] [1-8] 
castle ::= [O] [-] [O] castle_17 
move_9 ::= [+#] | 
nonpawn_10 ::= [a-h] | 
nonpawn_11 ::= [1-8] | 
nonpawn_12 ::= [x] | 
pawn_13 ::= [a-h] [x] 
pawn_14 ::= pawn_13 | 
pawn_15 ::= [=] [NBKQR] 
pawn_16 ::= pawn_15 | 
castle_17 ::= [-] [O] | 

 A good game:

1. e4 e5
2. Nf3 Nc6
3. Bb5 a6
4. Ba4 Nf6

llama_print_timings:        load time =  1144.33 ms
llama_print_timings:      sample time =    35.87 ms /    32 runs   (    1.12 ms per token)
llama_print_timings: prompt eval time =  1126.34 ms /     7 tokens (  160.91 ms per token)
llama_print_timings:        eval time =  5214.99 ms /    31 runs   (  168.23 ms per token)
llama_print_timings:       total time =  6398.45 ms
```

</details>

<details>
<summary>"Chess" without grammar</summary>

```
% ./main -m $LLAMA_30B_Q4_0 -n 32 -p $'A good game:\n\n'  

main: build = 645 (fd0eb66)
main: seed  = 1686286016
llama.cpp: loading model from /Users/evan/llama-models/30B/ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 6656
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 52
llama_model_load_internal: n_layer    = 60
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 17920
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 30B
llama_model_load_internal: ggml ctx size =    0.13 MB
llama_model_load_internal: mem required  = 19756.66 MB (+ 3124.00 MB per state)
.
llama_init_from_file: kv self size  =  780.00 MB

system_info: n_threads = 8 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | 
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.800000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 512, n_batch = 512, n_predict = 32, n_keep = 0


 A good game:

Sir Thomas Gresham, when he was building his famous Exchange at London, had the following dialogue with a mason, whose name was Richard B
llama_print_timings:        load time =  1185.47 ms
llama_print_timings:      sample time =    21.57 ms /    32 runs   (    0.67 ms per token)
llama_print_timings: prompt eval time =  1167.67 ms /     7 tokens (  166.81 ms per token)
llama_print_timings:        eval time =  4977.97 ms /    31 runs   (  160.58 ms per token)
llama_print_timings:       total time =  6188.21 ms
```

</details>

<details>
<summary>Arithmetic</summary>

```
 % ./main -m $LLAMA_30B_Q4_0 -n 32 -p $'Some arithmetic practice:\n\n' \                      
--grammar 'root  ::= (expr "=" ws num "\n")+
expr  ::= term ([-+*/] term)*
term  ::= ident | num | "(" ws expr ")" ws
ident ::= [a-z] [a-z0-9_]* ws
num   ::= [0-9]+ ws
ws    ::= [ \t\n]*'
main: build = 674 (e550234)
main: seed  = 1688014196
llama.cpp: loading model from /Users/evan/llama-models/30B/ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 6656
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 52
llama_model_load_internal: n_layer    = 60
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 17920
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 30B
llama_model_load_internal: ggml ctx size =    0.13 MB
llama_model_load_internal: mem required  = 19756.66 MB (+ 3124.00 MB per state)
.
llama_init_from_file: kv self size  =  780.00 MB

system_info: n_threads = 8 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | 
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.800000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 512, n_batch = 512, n_predict = 32, n_keep = 0


main: grammar:
root ::= root_5 
root_1 ::= expr [=] ws num [<U+000A>] 
expr ::= term expr_8 
ws ::= ws_12 
num ::= num_11 ws 
root_5 ::= root_1 root_5 | root_1 
term ::= ident | num | [(] ws expr [)] ws 
expr_7 ::= [-+*/] term 
expr_8 ::= expr_7 expr_8 | 
ident ::= [a-z] ident_10 ws 
ident_10 ::= [a-z0-9_] ident_10 | 
num_11 ::= [0-9] num_11 | [0-9] 
ws_12 ::= [ <U+0009><U+000A>] ws_12 | 

 Some arithmetic practice:

10 *a*1 +b*2 =640

10 *a*2 +b*3 =656


llama_print_timings:        load time =  1165.00 ms
llama_print_timings:      sample time =    41.11 ms /    32 runs   (    1.28 ms per token)
llama_print_timings: prompt eval time =  1147.76 ms /     7 tokens (  163.97 ms per token)
llama_print_timings:        eval time =  5113.92 ms /    31 runs   (  164.97 ms per token)
llama_print_timings:       total time =  6323.27 ms
```

</details>

<details>
<summary>Arithmetic - no grammar</summary>

```
 % ./main -m $LLAMA_30B_Q4_0 -n 32 -p $'Some arithmetic practice:\n\n'                                            
main: build = 645 (fd0eb66)
main: seed  = 1686286388
llama.cpp: loading model from /Users/evan/llama-models/30B/ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 6656
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 52
llama_model_load_internal: n_layer    = 60
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 17920
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 30B
llama_model_load_internal: ggml ctx size =    0.13 MB
llama_model_load_internal: mem required  = 19756.66 MB (+ 3124.00 MB per state)
.
llama_init_from_file: kv self size  =  780.00 MB

system_info: n_threads = 8 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | 
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.800000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 512, n_batch = 512, n_predict = 32, n_keep = 0


 Some arithmetic practice:

\begin{code}
package main

import (
    "fmt"
)

func main() {
    fmt.Println(
llama_print_timings:        load time =  1171.65 ms
llama_print_timings:      sample time =    21.37 ms /    32 runs   (    0.67 ms per token)
llama_print_timings: prompt eval time =  1153.88 ms /     7 tokens (  164.84 ms per token)
llama_print_timings:        eval time =  4991.68 ms /    31 runs   (  161.02 ms per token)
llama_print_timings:       total time =  6187.91 ms
```

</details>

<details>
<summary>JSON</summary>

```
% ./main -m $LLAMA_30B_Q4_0 -n 64 -p $'A bit about me:\n\n' --grammar-file grammars/json.gbnf
main: build = 674 (e550234)
main: seed  = 1688014289
llama.cpp: loading model from /Users/evan/llama-models/30B/ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 6656
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 52
llama_model_load_internal: n_layer    = 60
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 17920
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 30B
llama_model_load_internal: ggml ctx size =    0.13 MB
llama_model_load_internal: mem required  = 19756.66 MB (+ 3124.00 MB per state)
.
llama_init_from_file: kv self size  =  780.00 MB

system_info: n_threads = 8 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | 
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.800000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 512, n_batch = 512, n_predict = 64, n_keep = 0


main: grammar:
root ::= object 
object ::= [{] ws object_11 [}] 
value ::= object | array | string | number | boolean 
array ::= [[] ws array_15 []] 
string ::= ["] string_16 ["] ws 
number ::= number_17 number_18 ws 
boolean ::= boolean_19 ws 
ws ::= [ <U+0009><U+000A>] ws | 
object_8 ::= string [:] ws value object_10 
object_9 ::= [,] ws string [:] ws value 
object_10 ::= object_9 object_10 | 
object_11 ::= object_8 | 
array_12 ::= value array_14 
array_13 ::= [,] ws value 
array_14 ::= array_13 array_14 | 
array_15 ::= array_12 | 
string_16 ::= [ <U+0009>!#-[]-~] string_16 | 
number_17 ::= [-] | 
number_18 ::= [0-9] number_18 | [0-9] 
boolean_19 ::= [t] [r] [u] [e] | [f] [a] [l] [s] [e] 

 A bit about me:

{
	"fullName": "Ramon Rodriguez",
	"username": "ramon",
	"email": "ramon@mail.com",
	"phoneNumber": "+1234567890",
	"address": {
		
llama_print_timings:        load time =  1273.70 ms
llama_print_timings:      sample time =    82.93 ms /    64 runs   (    1.30 ms per token)
llama_print_timings: prompt eval time =  1256.36 ms /     8 tokens (  157.04 ms per token)
llama_print_timings:        eval time = 10432.05 ms /    63 runs   (  165.59 ms per token)
llama_print_timings:       total time = 11795.36 ms
```

</details>

<details>
<summary>"JSON" - no grammar</summary>

```
 % ./main -m $LLAMA_30B_Q4_0 -n 32 -p $'A bit about me:\n\n'                                                                          
main: build = 645 (fd0eb66)
main: seed  = 1686286615
llama.cpp: loading model from /Users/evan/llama-models/30B/ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 6656
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 52
llama_model_load_internal: n_layer    = 60
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 17920
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 30B
llama_model_load_internal: ggml ctx size =    0.13 MB
llama_model_load_internal: mem required  = 19756.66 MB (+ 3124.00 MB per state)
.
llama_init_from_file: kv self size  =  780.00 MB

system_info: n_threads = 8 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | 
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.800000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 512, n_batch = 512, n_predict = 32, n_keep = 0


 A bit about me:

A former teacher, now a full-time writer. I am the author of two novels: _The Man in the Moon_ and _The Riddle
llama_print_timings:        load time =  1291.32 ms
llama_print_timings:      sample time =    21.48 ms /    32 runs   (    0.67 ms per token)
llama_print_timings: prompt eval time =  1274.63 ms /     8 tokens (  159.33 ms per token)
llama_print_timings:        eval time =  4990.01 ms /    31 runs   (  160.97 ms per token)
llama_print_timings:       total time =  6306.01 ms
```

</details>

<details>
<summary>Japanese</summary>

```
 % ./main -m $LLAMA_30B_Q4_0 -n 32 -p $'Building a website can be done in 10 simple steps (from the original Japanese):\n\n' --grammar-file grammars/japanese.gbnf
main: build = 674 (e550234)
main: seed  = 1688013430
llama.cpp: loading model from /Users/evan/llama-models/30B/ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 6656
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 52
llama_model_load_internal: n_layer    = 60
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 17920
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 30B
llama_model_load_internal: ggml ctx size =    0.13 MB
llama_model_load_internal: mem required  = 19756.66 MB (+ 3124.00 MB per state)
.
llama_init_from_file: kv self size  =  780.00 MB

system_info: n_threads = 8 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | 
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.800000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 512, n_batch = 512, n_predict = 32, n_keep = 0


main: grammar:
root ::= root_2 root_5 
jp-char ::= hiragana | katakana | punctuation | cjk 
root_2 ::= jp-char root_2 | jp-char 
root_3 ::= [ <U+0009><U+000A>] root_4 
root_4 ::= jp-char root_4 | jp-char 
root_5 ::= root_3 root_5 | 
hiragana ::= [<U+3041>-<U+309F>] 
katakana ::= [<U+30A1>-<U+30FF>] 
punctuation ::= [<U+3001>-<U+303E>] 
cjk ::= [<U+4E00>-<U+9FFF>] 

 Building a website can be done in 10 simple steps (from the original Japanese):

一、目的は何なのか
二、お客さまを思い出して
三、お客さまのこと
llama_print_timings:        load time =  2957.19 ms
llama_print_timings:      sample time =    42.67 ms /    32 runs   (    1.33 ms per token)
llama_print_timings: prompt eval time =  2941.56 ms /    21 tokens (  140.07 ms per token)
llama_print_timings:        eval time =  5384.28 ms /    31 runs   (  173.69 ms per token)
llama_print_timings:       total time =  8387.61 ms
```

</details>

<details>
<summary>Japanese - no grammar</summary>

```
% ./main -m $LLAMA_30B_Q4_0 -n 32 -p $'Building a website can be done in 10 simple steps (from the original Japanese):\n\n' 
main: build = 674 (e550234)
main: seed  = 1688013483
llama.cpp: loading model from /Users/evan/llama-models/30B/ggml-model-q4_0.bin
llama_model_load_internal: format     = ggjt v3 (latest)
llama_model_load_internal: n_vocab    = 32000
llama_model_load_internal: n_ctx      = 512
llama_model_load_internal: n_embd     = 6656
llama_model_load_internal: n_mult     = 256
llama_model_load_internal: n_head     = 52
llama_model_load_internal: n_layer    = 60
llama_model_load_internal: n_rot      = 128
llama_model_load_internal: ftype      = 2 (mostly Q4_0)
llama_model_load_internal: n_ff       = 17920
llama_model_load_internal: n_parts    = 1
llama_model_load_internal: model size = 30B
llama_model_load_internal: ggml ctx size =    0.13 MB
llama_model_load_internal: mem required  = 19756.66 MB (+ 3124.00 MB per state)
.
llama_init_from_file: kv self size  =  780.00 MB

system_info: n_threads = 8 / 12 | AVX = 0 | AVX2 = 0 | AVX512 = 0 | AVX512_VBMI = 0 | AVX512_VNNI = 0 | FMA = 0 | NEON = 1 | ARM_FMA = 1 | F16C = 0 | FP16_VA = 1 | WASM_SIMD = 0 | BLAS = 1 | SSE3 = 0 | VSX = 0 | 
sampling: repeat_last_n = 64, repeat_penalty = 1.100000, presence_penalty = 0.000000, frequency_penalty = 0.000000, top_k = 40, tfs_z = 1.000000, top_p = 0.950000, typical_p = 1.000000, temp = 0.800000, mirostat = 0, mirostat_lr = 0.100000, mirostat_ent = 5.000000
generate: n_ctx = 512, n_batch = 512, n_predict = 32, n_keep = 0


 Building a website can be done in 10 simple steps (from the original Japanese):

1. Determine your goal for your site.
2. Make a plan.
3. Select the domain name.
4. Choose web
llama_print_timings:        load time =  2955.05 ms
llama_print_timings:      sample time =    22.96 ms /    32 runs   (    0.72 ms per token)
llama_print_timings: prompt eval time =  2937.10 ms /    21 tokens (  139.86 ms per token)
llama_print_timings:        eval time =  5032.41 ms /    31 runs   (  162.34 ms per token)
llama_print_timings:       total time =  8013.71 ms
```

</details>

## Approach

### Grammar API

The `llama` API accepts a data structure representing a context-free grammar over 32-bit code points:
```
    // grammar element type
    enum llama_gretype {
        // end of rule definition
        LLAMA_GRETYPE_END            = 0,

        // start of alternate definition for rule
        LLAMA_GRETYPE_ALT            = 1,

        // non-terminal element: reference to rule
        LLAMA_GRETYPE_RULE_REF       = 2,

        // terminal element: character (code point)
        LLAMA_GRETYPE_CHAR           = 3,

        // modifies a preceding LLAMA_GRETYPE_CHAR or LLAMA_GRETYPE_CHAR_ALT to
        // be an inclusive range ([a-z])
        LLAMA_GRETYPE_CHAR_RNG_UPPER = 4,

        // modifies a preceding LLAMA_GRETYPE_CHAR or
        // LLAMA_GRETYPE_CHAR_RNG_UPPER to add an alternate char to match ([ab], [a-zA])
        LLAMA_GRETYPE_CHAR_ALT       = 5,
    };

    typedef struct llama_grammar_element {
        enum llama_gretype type;
        uint32_t           value; // Unicode code point or rule ID
    } llama_grammar_element;

    LLAMA_API struct llama_grammar * llama_grammar_init(
            const llama_grammar_element ** rules,
                                 size_t    n_rules,
                                 size_t    start_rule_index);
```

### Sampling

The grammar sampling code models a nondeterministic pushdown automaton, maintaining N stacks for the possible parse states. Sampling a token is done in two steps: a sampling API that filters candidates to those that match one of the parse stacks (`llama_sample_grammar`) and adding the chose token to the grammar (`llama_grammar_accept_token`).

### Examples

Adds `--grammar` and `--grammar-file` arguments to `main` taking a simple extended BNF to constrain generations. The parser for this format is implemented in `examples/grammar-parser.{h,cpp}`:

```
// ... Supports character
// ranges, grouping, and repetition operators. As an example, a grammar for
// arithmetic might look like:
//
// root  ::= expr
// expr  ::= term ([-+*/] term)*
// term  ::= num | "(" space expr ")" space
// num   ::= [0-9]+ space
// space ::= [ \t\n]*
```
The `root` rule identifies the start of the grammar.

~## Caveats~

- ~the binary format makes the code harder to understand and more brittle~
- ~the grammar contemplates 16-bit chars but it's just being applied to the 8-bit UTF-8 chars in token strings currently~
- ~the 1-char lookahead sampling is probably biasing generations in a weird way; further investigation on quality of outputs is probably needed~

ejones committed:
llama, main : constrain sampling to grammar

howard0su commented:
Suggest taking a file as grammar parameter and put several examples like what we did for prompts (in .\prompts folder).

tobi commented:
Incredibly useful contribution. It's really amazing how much this simplifies many use cases. 

I agree that it would be better if the grammar came from a file. 

Two snags I hit while trying this out:

- it crashes with --prompt-cache
- any empty lines in the grammar cause a crash


Some additional thoughts:
- Would love to have the grammars support empty lines and comments
- I wonder if the grammar could be compiled into a tensor of state transitions and run on the GPU
- I wonder if there is an optimization where the next token is already known form the grammar we could skip the inference and just add it? In many types of grammars like json or html that could really speed up generation
- I think it's worth allowing to reference full tokens form the grammar. Maybe something like @

slaren commented:
Very nice! I am wondering what is the rationale for not including the parser in the llama.cpp API.  Without it, most downstream users will be forced to manually make a copy of the parser in their code to support the feature, which is not great.
Also for usability, I think it would be a good idea to keep a copy of the binary grammar in `llama_grammar`, rather than asking the users to keep the provided copy alive. The overhead would be minimal, and it would simplify the code of downstream users.

ejones committed:
allow loading grammar from file

ejones committed:
fix whitespace errors

ejones committed:
handle & print parser errors

ejones committed:
add comments to grammar syntax and allow newlines where unambiguous

ejones committed:
Merge remote-tracking branch 'refs/remotes/upstream/master' into grammar

ejones committed:
add missing include

ejones commented:
Thanks all! Just added support for grammar files (with examples) and updated the grammar syntax to add shell-style comments and allow empty lines between rules, as well as newlines inside parenthesized groups.

> it crashes with --prompt-cache

I wonder if that was #1699 ? If so, should be fixed now

> I wonder if the grammar could be compiled into a tensor of state transitions and run on the GPU

Sounds cool, I don't know enough about GPU programming to comment on that myself. The grammar participates in the sampling layer, and I'm not sure if that leverages the GPU currently.

> I wonder if there is an optimization where the next token is already known form the grammar we could skip the inference and just add it?

This is definitely possible. That said, AFAIK the token would still need to be evaluated, and that seems to be the bottleneck. Maybe the optimization comes in being able to batch eval strings of such tokens?

> I think it's worth allowing to reference full tokens form the grammar

Neat idea. Would that be more of an optimization or to reference tokens that can't be expressed textually?

> what is the rationale for not including the parser in the llama.cpp API.

Honestly, I was trying to reduce the changes to `llama.cpp` itself. Agree it would be more convenient in the API.

>  I think it would be a good idea to keep a copy of the binary grammar

Makes sense. I left that out of this round of changes - if it's desired to have the grammar parser in the llama API, this may naturally fit with that change.

bullno1 commented:
First, this is amazing work.

This makes me wonder whether the entire sampling API should be pulled into something like `llama_samplers` instead.
External samplers can evolve independently of the core API.

The existing functions can be kept for compatibility.
AFAIK, the only thing we need is to expose the RNG.
And even then, the existence of that inside a state/context is debatable.
The context window is already managed by user code so why not sampling?

This reminds me  a lot of: https://lmql.ai/.
There is also https://github.com/1rgs/jsonformer where the input is a json schema which is not always easy to express in BNF.

> AFAIK the token would still need to be evaluated

Would it though?
We just immediately add it to the context.
It is done manually in user code now.

> Maybe the optimization comes in being able to batch eval strings of such tokens?

AFAIK, that's the case.
The initial prompt and the user input are submitted in a large batch.
The inference loop just feed the single chosen token back until eos.

> The grammar participates in the sampling layer, and I'm not sure if that leverages the GPU currently.

The current sampling is CPU.

Green-Sky commented:
> This makes me wonder whether the entire sampling API should be pulled into something like llama_samplers instead.

one of the discussion points for adding more llm generic tooling back into ggml(repo) was moving the sampler there. but afaik nothing happened yet :)

ejones commented:
> There is also https://github.com/1rgs/jsonformer where the input is a json schema

Was planning to tackle this next. I've got it more or less working locally in a branch off of this, at least with the examples on jsonformer's README. It uses a Python script to generate a JSON BNF that conforms to the schema.

ejones committed:
support alternates in root rule

ejones committed:
fix bugs with empty token and EOS

ejones committed:
adjust JSON grammar

ejones committed:
remove swp file

howard0su commented on llama.h:
```diff
@@ -263,6 +289,9 @@ extern "C" {
     LLAMA_API void llama_sample_temperature(struct llama_context * ctx, llama_token_data_array * candidates, float temp);
 
+    /// @details Apply constraints from grammar
+    LLAMA_API void llama_sample_grammar(struct llama_context * ctx, llama_token_data_array * candidates, const struct llama_grammar * grammar);
```
Can we make llama_grammar as a structure with two callbacks? So the other implementation of it can support context aware state machine instead?

ejones replied:
Do you mean like, the caller would provide the implementation of `llama_grammar` (via callbacks), from which the llama API determines which tokens are valid?

howard0su replied:
yes, so llama code will not assume the grammar implementation.

ejones replied:
Yeah, I'm open to that idea, assuming the grammar interface itself generalizes well to other implementations. I kind of designed this with the specific implementation in mind so that's not a guarantee.

ggerganov commented:
Great stuff!

I'm still wrapping my head around this.

- Yes, this can become part of a `llama.cpp` or `ggml` sampling API, but I guess for now we can keep it as example and see what are the pros and cons and learn how to use it most efficiently
- What happens when then next `N > 1` tokens are uniquely determined by the grammar? I guess we will sample them one by one, correct? What would it take to make it so that they are submitted to be processed as a batch? This would significantly speed up the inference in such cases

ejones commented:
> * Yes, this can become part of a `llama.cpp` or `ggml` sampling API, but I guess for now we can keep it as example and see what are the pros and cons and learn how to use it most efficiently

To clarify, this PR adds the core sampling functionality in `llama.cpp`, leaving the grammar parser out in examples. Should that all be moved to examples or just left as is?

> * What happens when then next `N > 1` tokens are uniquely determined by the grammar? I guess we will sample them one by one, correct? What would it take to make it so that they are submitted to be processed as a batch? This would significantly speed up the inference in such cases

Yes, that's correct. I think that's doable, I can take a stab at that.

SlyEcho commented:
> the grammar contemplates 16-bit chars but it's just being applied to the 8-bit UTF-8 chars in token strings currently

I don't understand this part. So it is converting to UTF-16? 

Another option would be to use token values but it will be more limiting.

SlyEcho commented on llama.cpp:
```diff
@@ -2112,6 +2280,32 @@ void llama_sample_frequency_and_presence_penalties(struct llama_context * ctx, l
+
+        // prune tokens based on first char only - in `llama_grammar_accept_token` we will find the
+        // full matching prefix of the selected token
+        const bool valid = str[0] == ' '
+            ? llama_grammar_peek(stacks_after_space, str[1])
+            : str[0] || id == eos
+            ? llama_grammar_peek(grammar->stacks,    id == eos ? 0 : str[0])
+            : false;
```
Is it possible to rewrite this to be clearer without the ternary ops?

Something like this, if it is indeed how I understood it?

```suggestion
        bool valid = false;
        if (id == eos) {
            valid = llama_grammar_peek(grammar->stacks, 0);
        } else if (str[0] == ' ') {
            valid = llama_grammar_peek(stacks_after_space, str[1]);
        } else if (str[0] != 0) {
            valid = llama_grammar_peek(grammar->stacks, str[0]);
        }
```

ejones replied:
Yes, that's better, thanks!

SlyEcho commented on examples/grammar-parser.cpp:
```diff
@@ -0,0 +1,330 @@
+
+                    // each char of a literal is encoded as a "range" of char - char
+                    outbuf.push_back(2);
+                    outbuf.push_back(char_pair.first);
+                    outbuf.push_back(char_pair.first);
```
It seems like it could be possible pack two bytes into a single `uint16_t` value? `hi << 8 | low` in little endian order.

SlyEcho replied:
There is also the issue here that since you're reading from `char` and not `unsigned char` or `uint8_t`, any byte over 0x7f might become sign extended as negative integers when cast into 16 bits.

ejones replied:
Yeah, my thinking was to eventually treat the char values as UTF-16 code units, but punted on Unicode support for now. 

And yes, `parse_char` does a conversion from `char` in the input string to `uint16_t`. I was maybe thinking to limit the BNF grammar to ascii and offer unicode escapes for higher values.

SlyEcho replied:
In that case there is not much point to using UTF-16, I would go all the way to UTF-32.

UTF-16 still has characters spanning multiple code units.

ejones replied:
Yes, but at worst you'd have a surrogate pair for non-BMP. I felt like that was a good tradeoff for per-char space usage, but can go to UTF-32 if we want.

ivanstepanovftw commented:
Just dont use repeat penalties to get best grammar as llama can

burke commented:
FWIW I'm adapting this code into an analogous feature for models running on torch. In my implementation, I'm doing grammar enforcement logit masking on the GPU across the full token set before selecting candidates: https://github.com/Shopify/torch-grammar/blob/df23e354083c909c70120e256ed34036c93f6714/grammar_sampler.py#L232-L239. The same strategy would probably work here if anyone was super motivated to try it.

ggerganov commented on llama.h:
```diff
@@ -284,6 +313,9 @@ extern "C" {
     /// @details Randomly selects a token from the candidates based on their probabilities.
     LLAMA_API llama_token llama_sample_token(struct llama_context * ctx, llama_token_data_array * candidates);
 
+    /// @details Accepts the sampled token into the grammar, possibly transforming to a new token
+    LLAMA_API llama_token llama_grammar_accept_token(struct llama_context * ctx, struct llama_grammar * grammar, llama_token token);
+
```
In what cases the token is being transformed?

ejones replied:
When only a prefix of the sampled token's chars match the grammar. For example, if the grammar is expecting "}" followed by, say, whitespace or a comma, and you sample "}{", you'd get back the "}" token. This can happen because this implementation only peeks one char ahead for the sampling penalties.

ggerganov replied:
> This can happen because this implementation only peeks one char ahead for the sampling penalties.

I see. So it wouldn't happen if the sampling penalties were applied optimally according to the grammar, but for now it's much easier to have the implementation peek at just one char, correct?

ejones replied:
Yes, correct.

ejones committed:
rewrite ternary expressions

Co-authored-by: Henri Vasserman <henv@hot.ee>

ejones commented:
@SlyEcho
> The parser doesn't understand UTF-8 so it will create rules that don't match as the user expects.

Yeah, my rough plan for Unicode support was to store UTF-16 code units in the binary grammar, re-encoding the token strings as UTF-16 when sampling. And maybe limiting the BNF grammar to ascii with Unicode escapes. But I punted on Unicode support for this initial version.

ggerganov reviewed (approved):
I think this is a great first iteration for supporting grammar-guided sampling.
We can iterate on it from `master`

@ejones 
Merge it when you think you are ready with the reviews

SlyEcho commented:
So, this grammar feature works in two parts:

1. Parse the grammar into a bytecode grammar program.
2. During generation check the grammar using a state machine.

My question is about this intermediate bytecode (or wordcode?) representation. Is there some performance benefit for it? Is the text grammar very hard to parse so it needs to be in binary form?

Would it not be easier to parse into and use some kind of graph data structures using C/C++ structs?

ejones commented:
>  Is there some performance benefit for it?
> Would it not be easier to parse into and use some kind of graph data structures using C/C++ structs?

Yes, I agree, it would be simpler and clearer to use an AST expressed in structs. I can make that adjustment.

I went with the binary format for a couple reasons: first, I hoped it would have good cache locality, although I admit I haven't benchmarked that or anything, and this could be achieved with structs as well. Second, I wanted to reduce the overall size of changes to llama.cpp itself, and keep the C API simple. I wasn't sure what the tolerance for changes there would be. This included keeping the parser external, meaning the AST/parser output is an input to the C API.

ejones committed:
use struct for grammar elements and add Unicode support

ejones commented:
Update on this: working on refactoring this to store the parsed grammar as structs. Also trying to think through the Unicode handling a bit more.

mattpulver commented:
Thank you for this useful addition. I've attempted to transform the [BNF grammar for SQL-2003 ](https://github.com/ronsavage/SQL) into a compatible format and read it in with this branch. After around a stack depth of 58k+ of `llama_grammar_advance_stack()` calling itself recursively I'm hitting a segfault when `malloc()` is called. I'm still continuing to debug this but just thought I'd share it in the meantime in case it is a useful test case, at least in terms of testing a large grammar file.

Fyi the attached .gbnf file is not production-ready in terms of representing valid SQL grammar, as many of the less-standard tokens are simply marked `"TODO"` for now. (The plan is to verify its size is compatible with this branch first before cleaning it up.)

[sql-2003-2.gbnf.gz](https://github.com/ggerganov/llama.cpp/files/11860442/sql-2003-2.gbnf.gz)

ejones commented:
Thanks @mattpulver !

zakkor commented:
I've been playing with this for the past week and it works splendidly. Truly a game-changer for building robust LLM applications. Thanks for the great work @ejones!

tucnak commented:
@ejones What unicode library do you have in mind? I'm not exactly sure what token space looks like when it comes to non-English texts but I would at least expect some kind of normalisation (in Unicode sense) or clusters would have to be properly supported at least for some grammars. I'm currently trying to work out a GPT-4 based solution that will hopefully be able to generate highly-specific (overfitted, if you will) grammars to guide generation on the fly, & funnily enough a great deal of it is in Ukrainian and sometimes a great deal of the actual grammar would represent specific combinations in the local, and more generic in the global. Ukrainian (cyrillic) is fairly straightforward in the EGC sense but other scripts like Arabic, which is apparently getting more representation due to https://noor.tii.ae/ and similar projects.

ejones commented:
@SlyEcho updated to add a bit more structure to the in-memory grammar. Kept it somewhat flat still (each rule is an array of atomic elements) for memory locality and simplicity in the C API. Let me know what you think. Also added UTF-8 decoding such that the "characters" of the grammar are now code points. It should now support Unicode in the grammar files and in generations.

mattpulver commented:
Re: SQL grammar support
Your recent commit f8baad2 added useful diagnostic info to identify malformed/undefined rules which existed in the above attachment `sql-2003-2.gbnf.gz`. This helped me to create an improved [sql-92.gbnf](https://github.com/ggerganov/llama.cpp/files/11907640/sql-92.gbnf.gz) file that is both smaller and passes the improved grammar checking.

After making the fixes I'm still getting a similar SEGFAULT as before:
```
$ gdb --args bin/main --grammar-file sql-92.gbnf -n32 -p "How many states are in the United States?\n"
...
high ::= [2] | [H] [i] [g] [h] left-paren [2] right-paren 
root ::= direct-SQL-statement 

Program received signal SIGSEGV, Segmentation fault.
0x00007ffff7aab70d in ?? () from /usr/lib/libc.so.6
(gdb) bt
#0  0x00007ffff7aab70d in ?? () from /usr/lib/libc.so.6
#1  0x00007ffff7aace93 in ?? () from /usr/lib/libc.so.6
#2  0x00007ffff7aad82a in malloc () from /usr/lib/libc.so.6
#3  0x00007ffff7cb089d in operator new (sz=310240) at /usr/src/debug/gcc/gcc/libstdc++-v3/libsupc++/new_op.cc:50
#4  0x000055555557b4c4 in std::__new_allocator<llama_grammar_element const*>::allocate (this=0x7fffff7ff300, __n=38780)
    at /usr/include/c++/13.1.1/bits/new_allocator.h:147
#5  0x000055555557a3f3 in std::allocator_traits<std::allocator<llama_grammar_element const*> >::allocate (__n=38780, 
    __a=...) at /usr/include/c++/13.1.1/bits/alloc_traits.h:482
#6  std::_Vector_base<llama_grammar_element const*, std::allocator<llama_grammar_element const*> >::_M_allocate (
    this=0x7fffff7ff300, __n=38780) at /usr/include/c++/13.1.1/bits/stl_vector.h:378
#7  0x000055555559ee32 in std::vector<llama_grammar_element const*, std::allocator<llama_grammar_element const*> >::_M_realloc_insert<llama_grammar_element const* const&> (this=0x7fffff7ff300, __position=0x7ffff7bf23c0)
    at /usr/include/c++/13.1.1/bits/vector.tcc:459
#8  0x0000555555596f70 in std::vector<llama_grammar_element const*, std::allocator<llama_grammar_element const*> >::push_back (this=0x7fffff7ff300, __x=@0x7fffff7ff2d8: 0x5555559829e0) at /usr/include/c++/13.1.1/bits/stl_vector.h:1289
#9  0x0000555555580df8 in llama_grammar_advance_stack (rules=std::vector of length 1351, capacity 1351 = {...}, 
    stack=std::vector of length 19391, capacity 38780 = {...}, 
    new_stacks=std::vector of length 77557, capacity 131072 = {...})
    at ejones/llama.cpp/llama.cpp:1900
#10 0x0000555555580e0f in llama_grammar_advance_stack (rules=std::vector of length 1351, capacity 1351 = {...}, 
    stack=std::vector of length 19391, capacity 38780 = {...}, 
    new_stacks=std::vector of length 77557, capacity 131072 = {...})
    at  ejones/llama.cpp/llama.cpp:1902
#11 0x0000555555580e0f in llama_grammar_advance_stack (rules=std::vector of length 1351, capacity 1351 = {...}, 
    stack=std::vector of length 19391, capacity 38780 = {...}, 
    new_stacks=std::vector of length 77557, capacity 131072 = {...})
    at  ejones/llama.cpp/llama.cpp:1902
...
#58189 0x0000555555580e0f in llama_grammar_advance_stack (rules=std::vector of length 1351, capacity 1351 = {...}, 
    stack=std::vector of length 1, capacity 1 = {...}, new_stacks=std::vector of length 77557, capacity 131072 = {...})
    at ejones/llama.cpp/llama.cpp:1902
#58190 0x0000555555580e0f in llama_grammar_advance_stack (rules=std::vector of length 1351, capacity 1351 = {...}, 
    stack=std::vector of length 1, capacity 1 = {...}, new_stacks=std::vector of length 77557, capacity 131072 = {...})
    at ejones/llama.cpp/llama.cpp:1902
#58191 0x000055555558150b in llama_grammar_init (rules=0x55555598f560, n_rules=1351, start_rule_index=1350)
    at ejones/llama.cpp/llama.cpp:2043
#58192 0x000055555555c188 in main (argc=9, argv=0x7fffffffe578)
    at ejones/llama.cpp/examples/main/main.cpp:311
```
It appears as if the `new_stacks` in `llama_grammar_advance_stack()` is growing as O(n^2) with each recursive iteration which my be a problem for large grammar files like this one. Roughly speaking: is it possible to navigate through the stack tree without having to materialize each step?

SlyEcho commented:
Gonna look into this soon.

ejones commented:
@mattpulver I reproduced the segfault and it appears the problem in this case is [left-recursive rules](https://en.wikipedia.org/wiki/Left_recursion) like `query-expression` -> `non-join-query-expression` -> `query-expression` or `query-term` -> `non-join-query-term` -> `query-term`. Since the grammar is processed top-down (and there's no special handling of left recursion), it's infinitely recursing on expanding these initial references.

A workaround would be to adjust the rules to eliminate left-recursion. E.g., maybe something along the lines of this?
```
query-expression ::= query-term (("UNION" | "EXCEPT") "ALL"? corresponding-spec? query-term)* |
                   ... (other cases if needed)
```

We could probably at least detect left recursion and error out early. Longer term, we could look into an implementation that would support such grammars.

ejones commented:
@tucnak at this point the grammars are defined over code points, with no specific handling or recognition of grapheme clusters. A particular grammar could recognize grapheme clusters (and in effect normalize them), and maybe that's what you're working towards with your solution?

mudler commented:
JFYI After playing with it a bit more, I've bumped into this while trying on ARM64+CUDA:

```
Jul 04 20:37:13 localhost local-ai[34380]: LLAMA_ASSERT: /usr/local/LocalAI/go-llama/llama.cpp/llama.cpp:2479: !new_stacks.empty()                                                                                   
Jul 04 20:37:13 localhost local-ai[34380]: SIGABRT: abort     
```

ejones committed:
Merge remote-tracking branch 'upstream/master' into grammar

ejones commented:
@mudler thanks for the feedback! Re: OpenAI functions, I also have a draft up at #1887 with a script to convert JSON schemas to grammars. Re: that assertion, that is triggered when the sampled token doesn't match the grammar at all. If you do run into it on llama.cpp, let me know the inputs and happy to look into it :).

mudler commented:
> @mudler thanks for the feedback! Re: OpenAI functions, I also have a draft up at #1887 with a script to convert JSON schemas to grammars. Re: that assertion, that is triggered when the sampled token doesn't match the grammar at all. If you do run into it on llama.cpp, let me know the inputs and happy to look into it :).

yes, awesome job! I've looked at that PR indeed, and slightly adapted to Golang to generate grammars directly from the requests - my first attempts where more simple though, with a chain of let first choose an action -> and then fill the params.

However, what I'm seeing is quite weird - it doesn't happen when using llama.cpp directly, but only when using it with the golang bindings (https://github.com/go-skynet/go-llama.cpp) on a particular setup I have (it just happens on ARM, on my x86_64 machine just runs fine). The same grammar (basically it's equivalent output from your sample in #1887 ) works fine on x86_64 but crashes with ARM+CUDA with the error above.

I'm suspecting something weird going on in the toolchain package combination (gcc/nvcc) - I've tried to trace it with gdb back with no luck so far, seems indeed that there is no match with the grammar rules (even if it does match on x86_64!).

Anyhow appreciated!  I don't want to distract you - it doesn't look an issue while running llama.cpp  directly.  I will collect more datapoints  and I'll try to figure out if it's something replicable or not.

KerfuffleV2 commented:
Perhaps consider using a more standard BNF syntax? For example: https://mdkrajnak.github.io/ebnftest/

The changes to existing grammars would be pretty minor:

```ebnf
root        ::= jp-char+ (#'[ \t\n]' jp-char+)*
jp-char     ::= hiragana | katakana | punctuation | cjk
hiragana    ::= #'[ぁ-ゟ]'
katakana    ::= #'[ァ-ヿ]'
punctuation ::= #'[、-〾]'
cjk         ::= #'[一-鿿]'
```

Looks like the only change really is quoting character classes like `#'[blah-blah]'` instead of using bare `[blah-blah]`. You don't have to support EBNF on your side, just using a similar format for the features you do support would make reuse of existing EBNF grammars a lot easier.

ejones committed:
add unicode escapes

ejones commented:
@KerfuffleV2 when digging into EBNF (basically on [Wikipedia](https://en.wikipedia.org/wiki/Extended_Backus%E2%80%93Naur_form)), I concluded that there isn't a clear, single standard EBNF to follow. I opted for a format that would match typical modern regex syntax. For this, [XML's EBNF](https://www.w3.org/TR/xml/#sec-notation) was a good starting point, but using `\x`/ `\u` / `\U` escapes in place of their `#x`.

In the project you linked, the `#'...'` syntax seems to be an artifact of Clojure:

> This project uses the [clojurescipt port](https://github.com/lbradstreet/instaparse-cljs) of [instaparse](https://github.com/Engelberg/instaparse) which provides an excellent EBNF parser with one quirk: regexs must be quoted as clojure regex literals. For example the regex for any character A-Z, [A-Z], is entered as #'[A-Z]'.

I like the idea of reusing existing grammars. Perhaps it makes sense to implement alternate parsers or translators for the most popular formats? For example, it looks like ANTLR publishes a [large set](https://github.com/antlr/grammars-v4) of example grammars, but I don't know that we'd want to tie our main format to ANTLR to benefit from that.

ejones commented:
@SlyEcho gentle nudge - any comments on this?

ejones committed:
Merge remote-tracking branch 'upstream/master' into grammar

wanicca commented:
Hi, after reading the code I came up a question. 

Say we have a rule that the model must generate either `apple` or `banana`.

Current implementation will reserve all tokens with prefix `a` in the filtered candidates, such as `ant`, `ate`, `apple`.

So if `ant` is sampled, it will be truncated into `a`.

Eventually, we may get `a,p,p,l,e` (five tokens) rather than `apple` (one token).

Did I mistake something?

SlyEcho commented:
> @SlyEcho gentle nudge - any comments on this?

Sorry, I haven’t found time to test it yet but it looks pretty good already.

ejones commented:
> Eventually, we may get a,p,p,l,e (five tokens) rather than apple (one token).

ejones commented:
@wanicca 
> Eventually, we may get a,p,p,l,e (five tokens) rather than apple (one token).

Correct. Although even with this 1-char implementation, in cases where one or more longer tokens are uniquely determined by the grammar (e.g., "a" implies "apple"), the batch sampling optimization suggested by @ggerganov may also fix this.

SlyEcho commented:
I suppose we could have backtracking sometime in the future?

KerfuffleV2 commented:
> In the project you linked, the #'...' syntax seems to be an artifact of Clojure:

Ah, fair enough. It seems like the Rust `enbf` crate conforms to that ENBF tester for what it's worth.

> Although even with this 1-char implementation, in cases where one or more longer tokens are uniquely determined by the grammar [...]

Isn't this more of a problem than it might seem at first glance because models are trained with certain combinations of tokens? So causing the model to generate `a`,`p`,`p`,`l`,`e` piecemeal would likely cause the model to generate worse output and the reason why wouldn't really be apparent because it still just looks like "apple" to the user.

marclove commented:
> > In the project you linked, the #'...' syntax seems to be an artifact of Clojure:
> 
> Ah, fair enough. It seems like the Rust `enbf` crate conforms to that ENBF tester for what it's worth.
> 
> > Although even with this 1-char implementation, in cases where one or more longer tokens are uniquely determined by the grammar [...]
> 
> Isn't this more of a problem than it might seem at first glance because models are trained with certain combinations of tokens? So causing the model to generate `a`,`p`,`p`,`l`,`e` piecemeal would likely cause the model to generate worse output and the reason why wouldn't really be apparent because it still just looks like "apple" to the user.

If you're constraining your output based on a grammar, isn't violating the grammar the worst possible output? At least in the use cases I can think of, if I'm constraining on grammar, that's my first priority; qualitative measures separate from that are secondary.

Thanks for the work, @ejones. Looking forward to seeing this merged.

KerfuffleV2 commented:
> If you're constraining your output based on a grammar, isn't violating the grammar the worst possible output? 

It depends. That's certainly true sometimes, but unless the cons are clearly documented/apparent then users also might use grammar sampling to direct responses in the direction they prefer without necessarily understanding the tradeoff.

> if I'm constraining on grammar, that's my first priority; qualitative measures separate from that are secondary.

I'd say it depends again, even if sticking to the grammar is paramount you _still_ have to end up with a result with a quality level that's worth using for your application.

***

Anyway, I want to be clear, I'm not opposing this pull or saying it shouldn't be merged or anything like that. My motivation for bringing up that point is in interests of improving llama.cpp either by considering/exploring other approaches that may not involve these tradeoffs or ensuring they're documented when they exist.

I'm not just being negative/contrary.

ejones commented:
> causing the model to generate a,p,p,l,e piecemeal would likely cause the model to generate worse output and the reason why wouldn't really be apparent because it still just looks like "apple" to the user

Yes, I agree. I think the instructions and examples in the prompt can reduce the likelihood of this, but it is a problem. Maybe I should revisit the 1-char assumption.

ejones committed:
add inverse char ranges

ejones committed:
only sample full tokens (no peeking or truncation)

ejones commented:
@wanicca @KerfuffleV2 @ggerganov the latest update restricts sampling to complete token matches, removing the 1-char peeking and token truncation.

I noticed it's still possible for the model to inadvertently split up tokens, as it may select a prefix of another token where specific sequences are expected (e.g., selecting the "a" or "b" tokens at "apple" | "banana"). There's probably a further improvement to select the longest matching token in such cases, or something.

ggerganov committed:
llama : minor style changes

blindly applied in online editor - hopefully I didn't break something

ggerganov reviewed (approved):
Great stuff! Looking forward to playing with this

Merge it when you wish

brandonwillard commented:
Looks like you're working on the same problem we were here: https://github.com/normal-computing/outlines/pull/131.  We put together a really rough draft a little while ago regarding an approach to pre-computing the supports of these vocabulary token distributions: https://arxiv.org/abs/2307.09702.  We currently have an implementation for the regular expression/FSM case in [`outlines`](https://github.com/normal-computing/outlines), and we're extending it to CFGs right now.  Would be happy to hear your thoughts, if you're interested.

ejones committed:
update help text

ejones committed:
add warning message if EOS is disabled

ejones committed:
Merge remote-tracking branch 'upstream/master' into grammar

ejones commented:
@mankeli it can work alongside `--ignore-eos` if you supply a grammar that itself is infinite and thus rules out the EOS token, e.g., `--grammar 'root ::= ("a" | "b" | "c") root'` (although `--ignore-eos` is redundant then). That said, I've added a warning message for this specific case if EOS is disabled to hopefully clarify failures.

ejones commented:
@brandonwillard nice, thanks for sharing that! I like the idea of a trie of parse states to vocabulary tokens. Will have to take a deeper look as we optimize this implementation. As for the current approach, the sampling is O(N) on the vocabulary and there are many other sub optimal behaviors, but the grammar sampling hasn't turned out to be a bottleneck (at least on my machine and with the cases I've tried so far). The other sampling functions, e.g., repetition penalty, are also generally O(N) on vocabulary so it's not out of the ordinary here. Regarding your discussion of parser tokens, the parser here does operate on characters so that can be applied directly to token strings.

ggerganov commented:
Adding a similar grammar-based sampling to `whisper.cpp` would be a really cool contribution.
Here is one ongoing implementation that can potentially benefit from such functionality by using VIM-based grammar:

https://github.com/ggerganov/whisper.cpp/pull/1144

x4080 commented:
Can we improve the result by fine tuning the model ? If so, what is the example to improve it ?
TIA

ejones commented:
@ggerganov agreed! Although I'm not sure when or if I'll be able to contribute that.

@x4080 this approach is independent of the model variant and can be used with a fine-tune. In the comments above there's a demonstration of using WizardLM, for example.

x4080 commented:
@ejones thanks

RevanthRameshkumar commented:
What is the generation speed on this? How does it compare to unconstrained generation when using cpp?

I am trying to replicate this in python using an A100 on an 8bit quantized llama 7b and it is extremely slow per token (compared to unconstrained generation) due to all the extra encoding and decoding that needs to happen.

ejones commented:
I've only done CPU inference, but the performance impact has been insignificant for everything I've tried. On the M2 Max I'm seeing about ~0.5 ms / token sampling for unconstrained vs ~6ms with a grammar, with token eval taking about ~70ms for 13b (Q4_K).

I'm not sure if I know enough about GPU programming to meaningfully comment; I know that other folks are working on approaches that are more generic or are GPU- and/or Python-focused. There are some examples of this upthread: https://github.com/Shopify/torch-grammar and https://github.com/normal-computing/outlines.

ejones commented:
@ggerganov alright, I've done it: https://github.com/ggerganov/whisper.cpp/pull/1229

kalomaze commented:
Is it a natural side effect of the grammar sampling method that I am seeing a significant degradation in tokens per second speed during generation?
On 13b, I go from ~20T/s to ~13T/s, which is pretty unfortunate. But is it expected? Is it CPU bound potentially?

Green-Sky commented:
It tells you the time spent on sampling at the end. You can confirm it there.
eg:
```
llama_print_timings:        load time =   285,92 ms
llama_print_timings:      sample time =    30,81 ms /    37 runs   (    0,83 ms per token,  1200,95 tokens per second)
llama_print_timings: prompt eval time =   159,23 ms /     5 tokens (   31,85 ms per token,    31,40 tokens per second)
llama_print_timings:        eval time =  3656,10 ms /    36 runs   (  101,56 ms per token,     9,85 tokens per second)
llama_print_timings:       total time =  3861,11 ms
```

ejones commented:
Yeah, I generally see about ~5ms/token overhead for grammars on the M2 Max, which is usually a fraction of the per token eval time. But recently I was testing with a grammar and saw a more significant impact. Should investigate, there may be some pathological cases.
